snapcraft/internal/project_loader/_extensions/gnome-3-28.py

```diff
@@ -0,0 +1,111 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+# Import types and tell flake8 to ignore the "unused" List.
+from typing import Any, Dict, List  # noqa: F401
+
+from snapcraft.internal import errors
+from ._desktop_common import DesktopCommonExtension
+
+
+class GnomeExtensionClassicConfinementError(errors.SnapcraftError):
+    fmt = "The gnome extension doesn't support classic confinement."
+
+
+class Gnome_3_28Extension(DesktopCommonExtension):
+    """The Gnome extension.
+    This extension is to be used by applications that require GTK+.
+    Examples might include productivity applications or utilities.
+    Note that this extension does not support classically-confined snaps at this time.
+    """
+
+    supported_bases = ("core18",)
+
+    def __init__(self, yaml_data: Dict[str, Any]) -> None:
+        """Create a new GnomeExtension.
+        Note that this extension does not support classic snaps.
+        :param dict yaml_data: Loaded snapcraft.yaml data.
+        """
+
+        super().__init__(yaml_data)
+
+        if yaml_data.get("confinement") == "classic":
+            raise GnomeExtensionClassicConfinementError()
+
+        platform_snap = ""
+        base = yaml_data.get("base")
+        after_dependencies = []  # type: List[str]
+        dependency_part = {}  # type: Dict[str, Any]
+        if base is not None:
+            if base == "core18":
+                platform_snap = "gnome-3-28-1804"
```

Comment:
What happens if it's not? Should it raise an error to make sure this won't go out of sync with supported_bases above, which I assume is also being enforced?

Reply:
yes, you're quite right. I do think adding an extra exception here would be a good idea. Thanks :-)

Reply:
I've added an exception when the check falls-through. I also moved the exception for classic confinement into the extension loader itself rather than as extension-specific code. It made sense to allow other extensions to have declarative support for denying classic confinement. These are both in 3f5ec4f2.